KeeperSdk/vault/SecretManagerExtensions.cs

```diff
@@ -660,5 +662,105 @@ public async Task DeleteSecretManagerClient(string applicationId, string deviceI
             await Auth.ExecuteAuthRest("vault/app_client_remove", rq);
             await GetSecretManagerApplication(application.Uid, true);
         }
+
+        public async Task ShareSecretsManagerApplicationWithUser(string applicationId, string userUid, bool unshare, SharedFolderOptions sharedFolderOptions = null, DateTimeOffset expiration = default)
+        {
+            if (!TryGetKeeperRecord(applicationId, out var record))
+            {
+                throw new KeeperInvalidParameter("ShareSecretsManagerApplicationWithUser", "applicationId", applicationId, "Application not found");
+            }
+            ApplicationRecord application = record as ApplicationRecord ?? throw new KeeperInvalidParameter("ShareSecretsManagerApplicationWithUser", "applicationId", applicationId, "Application not found");
+
+            var appInfoResponse = await GetAppInfo(application.Uid);
+            var appInfo = appInfoResponse.FirstOrDefault(x => x.AppRecordUid.ToByteArray().SequenceEqual(application.Uid.Base64UrlDecode()));
+
+            var (sharedRecordPermissions, sharedfolderUserOptions) = GetFolderAndRecordPermissions(sharedFolderOptions, unshare, expiration);
+
+            // process application first with the user
+            await HandleRecordShareWithUser(unshare, application.Uid, userUid, sharedRecordPermissions);
+            // This is throwing an erorr which is probably NoActiveSessionWIthUser
+            // await HandleFolderShareWithUser(unshare, application.Uid, userUid, sharedfolderUserOptions);
+
+            // process each folder which application covers with the user
+            await Task.WhenAll(
+                appInfo.Shares
+                    .Where(x => x.ShareType == ApplicationShareType.ShareTypeFolder)
+                    .Select(x => HandleFolderShareWithUser(unshare, x.SecretUid, userUid, sharedfolderUserOptions))
+            );
+
+            // process each record which application covers with the user
+            await Task.WhenAll(
+                appInfo.Shares
+                    .Where(x => x.ShareType == ApplicationShareType.ShareTypeRecord)
+                    .Select(x => HandleRecordShareWithUser(unshare, x.SecretUid.ToByteArray().Base64UrlEncode(), userUid, sharedRecordPermissions))
+            );
+
+            // sync down the changes
+            await SyncDown();
+        }
```

Comment:
Can we move the code to share folder and records to separate functions. This will make the code readable.